pkg/kubevirt/utils.go

```diff
@@ -138,36 +153,86 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template)
+	switch cloudInitType {
+	case cloudInitNoCloud:
+		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
+		if err != nil {
+			return nil, err
+		}
+		return template, nil
+	case cloudInitConfigDrive:
+		template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
+	}
+	if !detectCloudInitDisk(template) {
+		template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+	}
+
+	return template, nil
+}
+
+// nodeRole returns the role of this node ("control-plane" or "worker").
+func nodeRole(ctx *context.MachineContext) string {
+	if util.IsControlPlaneMachine(ctx.Machine) {
+		return constants.ControlPlaneNodeRoleValue
+	}
+	return constants.WorkerNodeRoleValue
+}
+
+func cloudinitVolume(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, ctx *context.MachineContext, preferredType cloudInitType) kubevirtv1.Volume {
+	switch preferredType {
+	case cloudInitNoCloud:
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
 				},
 			},
-		},
+		}
+	default:
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
+				},
+			},
+		}
 	}
-	template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
+}
 
-	cloudInitDisk := kubevirtv1.Disk{
+func cloudinitDisk(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec) kubevirtv1.Disk {
+	return kubevirtv1.Disk{
 		Name: cloudInitVolumeName,
 		DiskDevice: kubevirtv1.DiskDevice{
 			Disk: &kubevirtv1.DiskTarget{
 				Bus: "virtio",
 			},
 		},
 	}
-	template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+}
 
-	return template
+func detectCloudInitVolumeType(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec) (preferredCloudInitVolumeType cloudInitType, existingCloudInitVolumeIndex int) {
+	for i, v := range vmi.Spec.Volumes {
+		if v.CloudInitNoCloud != nil && v.Name == cloudInitVolumeName {
+			return cloudInitNoCloud, i
+		}
+	}
+	return cloudInitConfigDrive, -1
```

Comment:
someone could theoretically set a `CloudInitConfigDrive` volume too. I don't like that this logic is assuming that only a CloudInitNoCloud volume get's the special behavior.

making the default cloudInitConfigDrive is good though

Reply:
this was the way to provide difference, since there's only two types implement currently.
the logic is extend able to show that other cloudInit types are found. it can be adapted when those are implement